src/compute-client/src/service.rs

```diff
@@ -133,13 +133,30 @@ pub struct PartitionedComputeState<T> {
     /// property ensures that a) we can eventually drop the tracking state maintained for a peek
     /// and b) we won't re-initialize tracking for a peek we have already served.
     peek_responses: BTreeMap<Uuid, BTreeMap<usize, PeekResponse>>,
-    /// Tracks in-progress `SUBSCRIBE`s, and the stashed rows we are holding
-    /// back until their timestamps are complete.
+    /// Tracks in-progress `SUBSCRIBE`s, and the stashed rows we are holding back until their
+    /// timestamps are complete.
     ///
     /// The updates may be `Err` if any of the batches have reported an error, in which case the
     /// subscribe is permanently borked.
-    pending_subscribes:
-        BTreeMap<GlobalId, Option<(MutableAntichain<T>, Result<Vec<(T, Row, Diff)>, String>)>>,
+    ///
+    /// Tracking of a subscribe is initialized when the first `SubscribeResponse` for that
+    /// subscribe is received. Once all shards have emitted an "end-of-subscribe" response the
+    /// subscribe tracking state is dropped again.
```

Comment:
What happens when a replica crashes while we're waiting for a subscribe response?

Reply:
We discussed this synchronously, but for posterity: A `PartitionedComputeState` is always scoped to a replica. That's true for the controller-side and the cluster-side one, the latter being scoped to a process of a replica. When a replica crashes, all its `PartitionedComputeState`s either crash as well (for cluster-side states) or get dropped by the controller (for the controller-side state). The controller rehydrates a crashed replica by dropping all its state and re-creating it. So the `PartitionedComputeState`s are re-created and tracking of subscribes (and all the other things) starts from the beginning again.

You might ask what exactly happens in the controller-side `PartitionedComputeState` when a process of a multi-process replica drops. In this case, `PartitionedComputeState` doesn't even notice. The `Partitioned` client (which is what actually implements `GenericClient` and calls `PartitionedComputeState` to perform compute-specific tasks) handles that in [its `recv` method](https://github.com/MaterializeInc/materialize/blob/ad971333440aefa5e2b35ec16600a2bb0e700707/src/service/src/client.rs#L135-L157) by returning an error to the caller. The caller is in this case the compute controller, which handles receive errors by marking the replica as failed an starting a rehydration for it.